### PR TITLE
adding pwdauth enhancements

### PR DESCRIPTION
**Pull request by cviecco**
Enhancing the pwlib to support the 2FA operations.

### CODE REVIEW

**Comment by rgooch** [comment]
Meta-comment before I dig into details: this doesn't seem to be the right place for Okta 2FA support. The lib/pwauth packages are clearly just for primary (password) authentication. Why not put this in a separate package for Okta 2FA?

**Comment by cviecco** [comment]
Because the okta2FA works as a "continuation" of the password auth. In fact it cannot live stand-alone. Thus the sensitive thing (for me) was to enhance the pwauth okta library to include the 2fa options. An option would be to move it to another place.

**Comment by rgooch** [comment]
Does Okta 2FA not have the option of just performing a 2nd factor check (i.e. no password required)?

**Comment by cviecco** [comment]
After rereading the docs, this is possible but only with trusted authentication as you need to create a user "authentication context" before doing the 2fa (you need to know what 2fa options your client has and then issue appropiate calls for each of the mfa options).

File: lib/pwauth/okta/api.go [hunk 1]
```diff
@@ -31,3 +50,19 @@ func (pa *PasswordAuthenticator) PasswordAuthenticate(username string,
...
+func (pa *PasswordAuthenticator) ValidateUserOTP(authUser string, otpValue int) (bool, error) {
+	return pa.validateUserOTP(authUser, otpValue)
+}
+
+// Initialize and verify Push
+func (pa *PasswordAuthenticator) ValidateUserPush(authUser string) (PushResponse, error) {
+	return pa.validateUserPush(authUser)
+}
+
+// SetAuthnURL. For testing only, update the internal authURL so that the backend can be tested
```
Comment (rgooch) [inline lib/pwauth/okta/api.go]: Please unexport this method.

File: lib/pwauth/okta/api.go [hunk 2]
```diff
@@ -1,15 +1,34 @@
...
 
 import (
+	"sync"
+	"time"
+
 	"github.com/Cloud-Foundations/Dominator/lib/log"
 	"github.com/Cloud-Foundations/keymaster/lib/simplestorage"
 )
 
+type authCacheData struct {
```
Comment (rgooch) [inline lib/pwauth/okta/api.go]: Please unexport all fields.

File: lib/pwauth/okta/api.go [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
+	Response PrimaryResponseType
+	Expires  time.Time
+}
+
 type PasswordAuthenticator struct {
-	authnURL string
-	logger   log.Logger
+	authnURL   string
+	logger     log.Logger
+	Mutex      sync.Mutex
```
Comment (rgooch) [inline lib/pwauth/okta/api.go]: Please unexport this field.

File: lib/pwauth/okta/impl.go [hunk 4]
```diff
@@ -5,35 +5,74 @@ import (
...
 	"github.com/Cloud-Foundations/Dominator/lib/log"
 )
 
 const (
-	authPath           = "/api/v1/authn"
-	authEndpointFormat = "https://%s.okta.com" + authPath
+	authPath               = "/api/v1/authn"
+	authEndpointFormat     = "https://%s.okta.com" + authPath
+	factorsVerifyPathExtra = "/factors/%s/verify"
 )
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please unexport all types.
Please unexport all fields except those which need to be written by the JSON decoder.

File: lib/pwauth/okta/impl.go [hunk 5]
```diff
@@ -5,35 +5,74 @@ import (
...
 
-type loginDataType struct {
+type VerifyTOTPFactorDataType struct {
+	StateToken string `json:"stateToken,omitempty"`
+	PassCode   string `json:"passCode,omitempty"`
+}
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please convert all JSON names from camelCase to lower_case. This comment applies for all types.

File: lib/pwauth/okta/impl.go [hunk 6]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
 		return false, fmt.Errorf("bad status: %s", resp.Status)
 	}
 	decoder := json.NewDecoder(resp.Body)
-	var response responseType
+	var response PrimaryResponseType
 	if err := decoder.Decode(&response); err != nil {
 		return false, err
 	}
+	// TODO: change logger to debug capable
+	//pa.logger.Printf("oktaresponse=%+v", response)
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete commented code.

File: lib/pwauth/okta/impl.go [hunk 7]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+			continue
+		}
+		authURL := fmt.Sprintf(pa.authnURL+factorsVerifyPathExtra, factor.Id)
+		verifyStruct := VerifyTOTPFactorDataType{
+			StateToken: userResponse.StateToken,
+			PassCode:   fmt.Sprintf("%06d", otpValue),
+		}
+		// TODO: change logger type to allow debug logs
+		// pa.logger.Printf("AuthURL=%s", authURL)
+		// pa.logger.Printf("totpVerifyStruct=%+v", verifyStruct)
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete commented code.

File: lib/pwauth/okta/impl.go [hunk 8]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+		if !(factor.FactorType == "push" && factor.VendorName == "OKTA") {
+			continue
+		}
+		authURL := fmt.Sprintf(pa.authnURL+factorsVerifyPathExtra, factor.Id)
+		verifyStruct := VerifyTOTPFactorDataType{
+			StateToken: userResponse.StateToken,
+		}
+		// TODO update logger type to allow debug logs
+		// pa.logger.Printf("AuthURL=%s", authURL)
+		// pa.logger.Printf("totpVerifyStruct=%+v", verifyStruct)
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete commented code.

File: lib/pwauth/okta/impl.go [hunk 9]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+		}
+		switch response.FactorResult {
+		case "WAITING":
+			return PushResponseWaiting, nil
+		case "TIMEOUT":
+			return PushResonseTimeout, nil
+		default:
+			return PushResponseRejected, nil
+		}
+
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete blank line.

File: lib/pwauth/okta/impl.go [hunk 10]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+func (pa *PasswordAuthenticator) validateUserOTP(authUser string, otpValue int) (bool, error) {
+	userResponse, err := pa.getValidUserResponse(authUser)
+	if err != nil {
+		return false, err
+	}
+	if userResponse == nil {
+		return false, nil
+	}
+
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete blank line.

File: lib/pwauth/okta/impl.go [hunk 11]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+		var response PrimaryResponseType
+		if err := decoder.Decode(&response); err != nil {
+			return false, err
+		}
+		if response.Status != "SUCCESS" {
+			return false, nil
+		}
+		return true, nil
+	}
+
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: Please delete blank line.

File: lib/pwauth/okta/okta_test.go [hunk 12]
```diff
@@ -41,17 +44,88 @@ func authnHandler(w http.ResponseWriter, req *http.Request) {
...
+		encoder := json.NewEncoder(w)
+
+		if err := encoder.Encode(response); err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		return
+	case "push-send-invalidWrapper":
+		writeStatus(w, "INVALID")
+		return
+
```
Comment (rgooch) [inline lib/pwauth/okta/okta_test.go]: Please delete blank line.

File: lib/pwauth/okta/okta_test.go [hunk 13]
```diff
@@ -41,17 +44,88 @@ func authnHandler(w http.ResponseWriter, req *http.Request) {
...
+	default:
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+
```
Comment (rgooch) [inline lib/pwauth/okta/okta_test.go]: Please delete blank line.

File: lib/pwauth/okta/okta_test.go [hunk 14]
```diff
@@ -67,14 +141,41 @@ func setupServer() {
...
+	_, err = pa.ValidateUserPush("someuser")
+	if err != nil {
+		t.Fatal(err)
+	}
+	memStore := memstore.New()
+	err = pa.UpdateStorage(memStore)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = pa.SetAuthnURL("http://localhost.locanet")
```
Comment (rgooch) [inline lib/pwauth/okta/okta_test.go]: s/locanet/localnet/ ?

File: lib/pwauth/okta/okta_test.go [hunk 15]
```diff
@@ -118,11 +232,304 @@ func TestMfaRequired(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	if !valid {
+		t.Fatal("should have succeeded with good  user")
+	}
+}
+
+func TestMfaPushNonExisting(t *testing.T) {
+	setupServer()
+
```
Comment (rgooch) [inline lib/pwauth/okta/okta_test.go]: Please delete blank line.

File: lib/pwauth/okta/api.go [hunk 16]
```diff
@@ -31,3 +50,19 @@ func (pa *PasswordAuthenticator) PasswordAuthenticate(username string,
 func (pa *PasswordAuthenticator) UpdateStorage(storage simplestorage.SimpleStore) error {
 	return nil
 }
+
+// VerifyOTP
+func (pa *PasswordAuthenticator) ValidateUserOTP(authUser string, otpValue int) (bool, error) {
```
Comment (rgooch) [inline lib/pwauth/okta/api.go]: s/authUser/username/

File: lib/pwauth/okta/api.go [hunk 17]
```diff
@@ -31,3 +50,19 @@ func (pa *PasswordAuthenticator) PasswordAuthenticate(username string,
...
+	return pa.validateUserOTP(authUser, otpValue)
+}
+
+// Initialize and verify Push
+func (pa *PasswordAuthenticator) ValidateUserPush(authUser string) (PushResponse, error) {
```
Comment (rgooch) [inline lib/pwauth/okta/api.go]: s/authUser/username/

File: lib/pwauth/okta/impl.go [hunk 18]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+		pa.Mutex.Lock()
+		pa.recentAuth[username] = toCache
+		pa.Mutex.Unlock()
 		return true, nil
 	default:
 		return false, nil
 	}
 }
+
+func (pa *PasswordAuthenticator) getValidUserResponse(authUser string) (*PrimaryResponseType, error) {
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: s/authUser/username/

File: lib/pwauth/okta/impl.go [hunk 19]
```diff
@@ -50,21 +89,173 @@ func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
...
+		pa.Mutex.Lock()
+		delete(pa.recentAuth, authUser)
+		pa.Mutex.Unlock()
+		return nil, nil
+
+	}
+	return &userData.Response, nil
+}
+
+func (pa *PasswordAuthenticator) validateUserOTP(authUser string, otpValue int) (bool, error) {
```
Comment (rgooch) [inline lib/pwauth/okta/impl.go]: s/authUser/username/

File: lib/pwauth/okta/api.go (hunk 1 above)
Comment (cviecco) [reply lib/pwauth/okta/api.go]: This will be needed by later PR so that we can do testing of the handler that calls these methods.

Comment (rgooch) [reply lib/pwauth/okta/api.go]: How about a new function that creates a testing *PasswordAuthenticator?

**Comment by rgooch** [comment]
As I think about it more, I think when this gets moved to the `golib` repository that we should drop these changes from this package and restore the original API, and instead create a 2FA package for Okta which requires an API token (just like SymantecVIP does).

Using the Okta public access API is a short-term solution as anyone in the world can DOS the API for your domain. Using an API token isolates you from that.

I also think that this is bending (breaking) our clean abstractions to suit the way Okta does things. That could just make things harder if someone wants to add support for a different provider.

**Comment by rgooch** [comment]
Ping? What's happening with this code?

File: lib/pwauth/okta/impl.go (hunk 5 above)
Comment (cviecco) [reply lib/pwauth/okta/impl.go]: This is the okta API... this cannot be changed.

**Comment by rgooch** [comment]
Reviewing the actual code change is tricky because the diff just shows new files created and old files deleted. Did you use `git mv` and then edit the files (in theory that should show up as a move/edit, which is better to review).

**Comment by cviecco** [comment]
I did a git mv.

File: lib/authenticators/okta/api.go [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/Cloud-Foundations/golib/pkg/log"
+	"github.com/Cloud-Foundations/keymaster/lib/simplestorage"
+)
+
+// This module implements the PasswordAuthenticator interface and will implement
+// a unified 2fa backend interface in some future
+
+type authCacheData struct {
+	Response PrimaryResponseType
+	Expires  time.Time
```
Comment (rgooch) [inline lib/authenticators/okta/api.go]: Please make these fields private.

File: lib/authenticators/okta/api.go [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+	return pa.validateUserOTP(username, otpValue)
+}
+
+// Initialize and verify Push
+func (pa *PasswordAuthenticator) ValidateUserPush(username string) (PushResponse, error) {
+	return pa.validateUserPush(username)
+}
+
+// SetAuthnURL. For testing only, update the internal authURL so that the backend can be tested
+func (pa *PasswordAuthenticator) SetAuthnURL(authnURL string) error {
```
Comment (rgooch) [inline lib/authenticators/okta/api.go]: Please replace this with a function that creates a testing object. It should not be possible to change a particular object from "normal" to "test".

File: lib/authenticators/okta/impl.go [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+	"github.com/Cloud-Foundations/golib/pkg/log"
+)
+
+const (
+	authPath               = "/api/v1/authn"
+	authEndpointFormat     = "https://%s.okta.com" + authPath
+	factorsVerifyPathExtra = "/factors/%s/verify"
+)
+
+type VerifyTOTPFactorDataType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+	StateToken string `json:"stateToken,omitempty"`
+	PassCode   string `json:"passCode,omitempty"`
+}
+
+type LoginDataType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+	Password string `json:"password,omitempty"`
+	Username string `json:"username,omitempty"`
+}
+
+type MFAFactorsType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+	Id         string `json:"id,omitempty"`
+	FactorType string `json:"factorType,omitempty"`
+	Provider   string `json:"provider,omitempty"`
+	VendorName string `json:"vendorName,omitempty"`
+}
+
+type UserProfileType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+	Login string `json:"login,omitempty"`
+}
+
+type UserInfoType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+	Id      string          `json:"id,omitempty"`
+	Profile UserProfileType `json:"profile,omitempty"`
+}
+
+type EmbeddedDataResponseType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 28]
```diff
@@ -0,0 +1,261 @@
...
+	User   UserInfoType     `json:"user,omitempty"`
+	Factor []MFAFactorsType `json:"factors,omitempty"`
+}
+
+type PrimaryResponseType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+	StateToken      string                   `json:"stateToken,omitempty"`
+	ExpiresAtString string                   `json:"expiresAt,omitempty"`
+	Status          string                   `json:"status,omitempty"`
+	Embedded        EmbeddedDataResponseType `json:"_embedded,omitempty"`
+}
+
+type PushResponseType struct {
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please unexport type.

File: lib/authenticators/okta/impl.go [hunk 30]
```diff
@@ -0,0 +1,261 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("bad status: %s", resp.Status)
+	}
+	decoder := json.NewDecoder(resp.Body)
+	var response PrimaryResponseType
+	if err := decoder.Decode(&response); err != nil {
+		return false, err
+	}
+	// TODO: change logger to debug capable
+	//pa.logger.Printf("oktaresponse=%+v", response)
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Please change the API to require a `DebugLogger`. The caller already passes one in.

File: lib/authenticators/okta/impl.go [hunk 31]
```diff
@@ -0,0 +1,261 @@
...
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+func (pa *PasswordAuthenticator) getValidUserResponse(username string) (*PrimaryResponseType, error) {
+	pa.Mutex.Lock()
+	userData, ok := pa.recentAuth[username]
+	pa.Mutex.Unlock()
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Unless you're doing slow/blocking work between the two locked sections, it's often faster to put all the code in a single lock/defer unlock.
Also, with two lock/unlock sections it's harder to reason whether this is safe.

File: lib/authenticators/okta/impl.go [hunk 32]
```diff
@@ -0,0 +1,261 @@
...
+			continue
+		}
+		authURL := fmt.Sprintf(pa.authnURL+factorsVerifyPathExtra, factor.Id)
+		verifyStruct := VerifyTOTPFactorDataType{
+			StateToken: userResponse.StateToken,
+			PassCode:   fmt.Sprintf("%06d", otpValue),
+		}
+		// TODO: change logger type to allow debug logs
+		// pa.logger.Printf("AuthURL=%s", authURL)
+		// pa.logger.Printf("totpVerifyStruct=%+v", verifyStruct)
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Same comment about switching to `DebugLogger`.

File: lib/authenticators/okta/impl.go [hunk 33]
```diff
@@ -0,0 +1,261 @@
...
+		if !(factor.FactorType == "push" && factor.VendorName == "OKTA") {
+			continue
+		}
+		authURL := fmt.Sprintf(pa.authnURL+factorsVerifyPathExtra, factor.Id)
+		verifyStruct := VerifyTOTPFactorDataType{
+			StateToken: userResponse.StateToken,
+		}
+		// TODO update logger type to allow debug logs
+		// pa.logger.Printf("AuthURL=%s", authURL)
+		// pa.logger.Printf("totpVerifyStruct=%+v", verifyStruct)
```
Comment (rgooch) [inline lib/authenticators/okta/impl.go]: Same comment about switching to `DebugLogger`.

File: lib/authenticators/okta/api.go [hunk 34]
```diff
@@ -64,8 +64,21 @@ func (pa *PasswordAuthenticator) ValidateUserPush(username string) (PushResponse
 	return pa.validateUserPush(username)
 }
 
+// New creates a new public authenticator, but poiting to an explicit authenticator url
```
Comment (rgooch) [inline lib/authenticators/okta/api.go]: s/poiting/pointing/

Also, please move this up in the file to before `NewPublic`.

File: lib/authenticators/okta/impl.go (hunk 30 above)
Comment (rgooch) [reply lib/authenticators/okta/impl.go]: Yes, please change the API.

File: lib/authenticators/okta/impl.go (hunk 31 above)
Comment (rgooch) [reply lib/authenticators/okta/impl.go]: Ping.

File: lib/pwauth/okta/okta_test.go (hunk 14 above)
Comment (rgooch) [reply lib/pwauth/okta/okta_test.go]: Ping.

File: lib/authenticators/okta/api.go (hunk 34 above)
Comment (cviecco) [reply lib/authenticators/okta/api.go]: fixed

File: lib/authenticators/okta/okta_test.go [hunk 35]
```diff
@@ -0,0 +1,533 @@
...
+	_, err = pa.ValidateUserPush("someuser")
+	if err != nil {
+		t.Fatal(err)
+	}
+	memStore := memstore.New()
+	err = pa.UpdateStorage(memStore)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = NewPublicTesting("http://localhost.locanet", testlogger.New(t))
```
Comment (rgooch) [inline lib/authenticators/okta/okta_test.go]: s/locanet/localnet/ ?

File: lib/authenticators/okta/api.go [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// Initialize and verify Push
+func (pa *PasswordAuthenticator) ValidateUserPush(username string) (PushResponse, error) {
+	return pa.validateUserPush(username)
+}
+
+// New creates a new public authenticator, but pointing to an explicit authenticator url
+func NewPublicTesting(authnURL string, logger log.DebugLogger) (
+	*PasswordAuthenticator, error) {
```
Comment (rgooch) [inline lib/authenticators/okta/api.go]: Please fix function name in comment.
Please move up in the file.

File: lib/authenticators/okta/api.go [hunk 37]
```diff
@@ -0,0 +1,76 @@
...
+
+const (
+	PushResponseRejected PushResponse = iota
+	PushResponseApproved
+	PushResponseWaiting
+	PushResonseTimeout
+)
+
+// New creates a new PasswordAuthenticator using Okta as the backend. The Okta
+// Public Application API is used, so rate limits apply.
```
Comment (rgooch) [inline lib/authenticators/okta/api.go]: Please fix function name in comment.

**Pull request merged**